engines/sci/util.h

```diff
@@ -45,6 +50,715 @@ void WRITE_SCI11ENDIAN_UINT32(void *ptr, uint32 val);
+		data = (typename RemoveConst<value_type>::type *)other.data;
+		_size = other._size;
+		_sourceName = other._sourceName;
+		_sourceByteOffset = other._sourceByteOffset;
+	}
+
+	template <typename OtherValueType>
+	inline Span &operator=(const Span<OtherValueType> &other) {
+		if (*this == other) {
+			return *this;
+		}
+
+		data = other.data;
+		_size = other._size;
+		_sourceName = other._sourceName;
+		_sourceByteOffset = other._sourceByteOffset;
+		return *this;
+	}
+
+	inline operator bool() const { return data != nullptr; }
+
+	template <typename OtherValueType>
+	inline bool operator==(const Span<OtherValueType> &other) const {
+		return data == other.data && _size == other._size;
+	}
+
+	/**
+	 * Creates a read stream of the data in memory.
+	 */
+	inline Common::MemoryReadStream toStream(const index_type index = 0, size_type dataSize = kSpanMaxSize) const {
+
+		if (dataSize == kSpanMaxSize) {
+			dataSize = _size - index;
+		}
+		validate(index, dataSize, kValidateSeek);
+		return Common::MemoryReadStream(data + index, dataSize * sizeof(value_type), DisposeAfterUse::NO);
+	}
+
+	inline size_type size() const { return _size; }
+	inline size_type byteSize() const { return _size * sizeof(value_type); }
+
+	inline const_iterator begin() const { return const_iterator(this, 0); }
+	inline iterator begin() { return iterator(this, 0); }
+	inline const_iterator end() const { return const_iterator(this, _size); }
+	inline iterator end() { return iterator(this, _size); }
+
+	inline const Common::String &getSourceName() const { return _sourceName; }
+	inline index_type getSourceByteOffset() const { return _sourceByteOffset; }
+
+	// Span are allowed to function as ForwardIterator for efficiency when
+	// looping over structures
+	inline Span &operator+=(const index_type delta) {
+		validate(delta, 0, kValidateSeek);
+		data += delta;
+		_size -= delta;
+		_sourceByteOffset += delta * sizeof(value_type);
+		return *this;
+	}
+
+	inline Span &operator++() {
+		return operator+=(1);
+	}
+
+	inline Span operator++(int) {
+		Span old = *this;
+		operator+=(1);
+		return old;
+	}
+
+	inline value_type operator*() {
+		validate(0, sizeof(value_type));
+		return data[0];
+	}
+
+	/**
+	 * Points the Span to null.
+	 */
+	inline void clear() {
+		data = nullptr;
+		_size = 0;
+		_sourceName = Common::String();
+		_sourceByteOffset = 0;
+	}
+
+	template <typename NewValueType>
+	inline const Span<NewValueType> subspan(const index_type index, const size_type newSize = kSpanMaxSize, const Common::String sourceName = Common::String()) const {
+		Span<NewValueType> span;
+		populateSubspan(span, index, newSize, sourceName);
+		return span;
+	}
+
+	template <typename NewValueType>
+	inline Span<NewValueType> subspan(const index_type index, const size_type newSize = kSpanMaxSize, const Common::String sourceName = Common::String()) {
+		return const_cast<Span &>(static_cast<const Span &>(*this).subspan(index, newSize, sourceName));
```

Comment:
I think duplicating the three lines from the const-version above is clearer than these casts.